File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2020/Y20D02.kt
package org.gristle.pdxpuzzles.advent.y2020

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y20D02(input: String) : Day {

    data class PassPolicy(val letter: Char, val range: IntRange, val password: String) {
        fun isValidUnderOldJobPolicy() = password.count { it == letter } in range
        fun isValidUnderCurrentPolicy() =
            (password[range.first - 1] == letter) xor (password[range.last - 1] == letter)
    }

    private val policies = input
        .lineSequence()
        .map {
            val (lower, upper, letter, password) = it.split('-', ' ')
            PassPolicy(letter.first(), lower.toInt()..upper.toInt(), password)
        }.toList()

    override fun part1() = policies.count(PassPolicy::isValidUnderOldJobPolicy)
    override fun part2() = policies.count(PassPolicy::isValidUnderCurrentPolicy)
}

fun main() = Day.runDay(Y20D02::class)

//    Class creation: 32ms
//    Part 1: 445 (0ms)
//    Part 2: 491 (0ms)
//    Total time: 33ms

File: rust/advent/src/bin/2020/y20d02.rs
use std::ops::{BitXor, RangeInclusive};
use itertools::Itertools;
use advent::utilities::get_input::get_input;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input = Vec<PassPolicy>;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(20, 2).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

struct PassPolicy {
    letter: char,
    range: RangeInclusive<usize>,
    password: String,
}

impl PassPolicy {
    fn valid_old(&self) -> bool {
        let count = self.password.chars().filter(|&c| c == self.letter).count();
        self.range.contains(&count)
    }
    fn valid_new(&self) -> bool {
        let a = self.password.as_bytes()[self.range.start() - 1] == self.letter as u8;
        let b = self.password.as_bytes()[self.range.end() - 1] == self.letter as u8;
        a.bitxor(b)
    }
}

fn parse_input(input: &str) -> Input {
    input.lines()
        .map(|line| {
            let (lower, upper, letter, password) = line
                .split(['-', ' '])
                .collect_tuple()
                .unwrap();
            PassPolicy {
                letter: letter.chars().next().unwrap(),
                range: lower.parse().unwrap()..=upper.parse().unwrap(),
                password: password.to_string(),
            }
        })
        .collect()
}

fn part1(policies: &Input) -> Output {
    policies.iter().filter(|p| p.valid_old()).count()
}

fn part2(policies: &Input) -> Output {
    policies.iter().filter(|p| p.valid_new()).count()
}

#[test]
fn default() {
    let input = get_input(20, 2).unwrap();
    let input = parse_input(&input);
    assert_eq!(445, part1(&input));
    assert_eq!(491, part2(&input));
}

// Input parsed (126μs)
// 1. 445 (18μs)
// 2. 491 (3μs)
// Total: 149μs